### DISCUSSION TITLE
React 이미지 경로 설정 (public vs. src)

### DISCUSSION THREAD

**Discussion by July249** [beyondDevelops/AnimalTalk#180] [General]
리액트에서 이미지 파일을 public 폴더에 저장하는지, 아니면 src 폴더 내에 저장하는지에 따라 각각의 장단점이 극명합니다.

우리는 이번 프로젝트에서 GitHub를 이용하여 배포를 진행할 예정이기 때문에 public 폴더로 관리를 하게 되었습니다. 

하지만, 각 경우에 대한 명확한 이해가 선행되어야 이를 사용한 의미가 있기 때문에 이에 대한 정리 글을 올립니다.

먼저 프로젝트 시작에 있어 해당 이슈를 설명해주신 재영님께 감사를 드립니다.

-----

## public 폴더에서 이미지 파일을 관리하는 경우의 특징

- webpack에서 관리를 해주지 않습니다. 대신 원본파일이 build 폴더에 복사됩니다. 즉 원본파일이 같이 올라갑니다.
- public 폴더에 접근하기 위해서 환경변수를 사용해야합니다. ( `process.env.PUBLIC_URL`)
- 경로 또는 파일이 없는 경우 컴파일 에러가 발생하지 않습니다. 단, 404 페이지 에러가 발생합니다.
- CRA 문서에서는 다음의 경우에 한하여 public 폴더 내 이미지 관리를 권고합니다.
  1. manifest, webmanifest 처럼 build된 결과물에서 특정 파일 이름이 필요한 경우
  2. 수 천 개의 이미지 파일을 동적으로 참조해야하는 경우
- GitHub에 소스코드를 올리는 경우 public 폴더에서 이미지를 관리 및 경로설정을 진행해야 정상적으로 이미지가 노출됨
- public 폴더를 root 경로로 잡음

## src 폴더에서 이미지 파일을 관리하는 경우의 특징

- 파일을 찾지 못한 경우 컴파일 에러가 발생합니다.
- import 시 참조할 수 있는 경로(path) 문자열을 출력하여 줍니다.
- content hash가 파일명에 포함되어 있기 때문에 브라우저가 오래된 캐싱을 가진 경우를 고려하지 않아도 됩니다. (단, 파일 변경시 hash도 변경됨)
- src 폴더를 root 경로로 잡음

**Comment by GreattitJY** [comment]
정리 감사합니다!
